WordPress/Classes/ViewRelated/Gutenberg/GutenbergSuggestionsViewController.swift

```diff
@@ -114,6 +114,23 @@ extension GutenbergSuggestionsViewController: UITextFieldDelegate {
             return true
         }
         let searchWord = nsString.replacingCharacters(in: range, with: string)
+
+        // This feels a bit hacky, so I'll explain what's being done here:
+        //     1. If the user types "@ ", the so-called "dismiss sequence",
+        //        the user probably typed the space to dismiss the suggestions UI.
+        //     2. So when this happens, we call the success handler and
+        //        pass in an empty string.
+        // We pass in an empty string because on the RN side here, the success
+        // handler inserts the @, followed by the username, followed by a space.
+        // Since we're essentially passing in an empty username, the result is
+        // that a "@ " is inserted into the post, which is the desired result.
+        // The same applies for cross-posts, with "+ " instead of "@ ".
+        let dismissSequence = suggestionType.trigger + " "
+        guard searchWord != dismissSequence else {
+            onCompletion?(.success(""))
```

Comment:
:wave: @guarani ! If I'm understanding this correctly, passing this as a "success" means that we'll [send off analytics indicating that the user successfully selected a suggestion](https://github.com/wordpress-mobile/WordPress-iOS/blob/bad64c922becf79ff52aa3126ef949ea9a6a40d6/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L914-L917), right? If that's what will happen, that doesn't feel right to me. Seems that we should be treating this as a failure case where the user did _not_ select a suggestion (at least for the purposes of the analytics we record).

I imagine that the reason you're using the success case is because we get the insertion of the "space" [for free](https://github.com/WordPress/gutenberg/blob/f1f63ab22da1994ab2a71056e092c8900f12467c/packages/rich-text/src/component/index.native.js#L495) when we do that, but maybe there are other reasons I'm not thinking of), whereas if we send a "failure" back to JS, then nothing gets inserted. It feels like we have two options here:

1. Send back a "success" to javascript (so the space gets inserted), but send a "failure" to the analytics (that's what @SiobhyB [has proposed doing on Android](https://github.com/wordpress-mobile/WordPress-Android/pull/14793/files)); or
2. Add a way to send text-to-be-inserted (in this case a single space) back to javascript with a "failure".

Option 2 feels a bit "cleaner" to me, but it may be over-engineering for the single case where we just want to send a space back since we already get that for free with a "success", so I'm sympathetic to just going with Option 1's simpler approach. If we do that though, I think we may need some comments explaining why we're doing this. Curious to hear what you (and @SiobhyB ) think though.

Reply:
Thanks for catching this, @mchowning! I completely overlooked the Tracks aspect. I'm going to revert this PR for the time-being while a fix is made. 

I think option 1 might work on iOS as well, but will take a look at option 2 as well.

Reply:
I created a revert PR [here](https://github.com/wordpress-mobile/WordPress-iOS/pull/16641).

Reply:
> Are there key differences to how the JavaScript side would handle a "success" vs a "failure" message?

Good question. The only difference that I can imagine at this point is the fact that we want to add a space at the end of a successfully inserted mention, whereas we might not want to do that when a mention was not inserted. For example, _if_ we wanted to have a space exit the mention at any point in the mentions UI (which as Paul points out, [we probably do not want to do](https://github.com/wordpress-mobile/WordPress-iOS/pull/16601#issuecomment-854121540)), then we wouldn't want to add an "extra" space on after the "mention text" which already ends in a space. 

This is a bit of a made up example though since we don't actually want this behavior, and this is my concern with Option 2: we don't really know what other use case there might be that would need failure-with-text, so it's hard to know how to best design it. 

Not that I'm opposed to Option 2--I bias pretty strongly to finding the cleanest approach most often--but in this case I'm not sure whether it's worth much extra effort: your implementation of Option 1 on Android is quite simple and works perfectly. Doing option 1 on iOS probably won't be quite so simple, so I'll be curious to see what Paul thinks after he digs in a bit.

Reply:
Ah, I see, thank you for that extra context!

Reply:
I left a comment, https://github.com/wordpress-mobile/gutenberg-mobile/issues/2943. Once that's resolved I think we'll have a better picture of how to handle this.

Reply:
> I left a comment, [wordpress-mobile/gutenberg-mobile#2943](https://github.com/wordpress-mobile/gutenberg-mobile/issues/2943). Once that's resolved I think we'll have a better picture of how to handle this.

Now that https://github.com/wordpress-mobile/gutenberg-mobile/issues/2943#issuecomment-869702797 is closed, we don't have another use case for designing a way to pass text back to the React Native world in the suggestions insertion "failure" case. I think [Option 1](https://github.com/wordpress-mobile/WordPress-iOS/pull/16601#discussion_r646606449), where we send a success event to React Native and a "failure" to analytics, is a reasonable solution here.

I'll go ahead with this on the iOS side and ping you @SiobhyB once this is ready just to sync up on merging these changes.

Reply:
Perfect, thank you @guarani!

Reply:
FIY the iOS PR is up: https://github.com/wordpress-mobile/WordPress-iOS/pull/16770